File: otto-mobile/otto-mobile/ChatReadCursorStore.swift
import Foundation

enum ChatReadThreadKey: Hashable, Codable {
    case squad(String)
    case direct(String)

    var storageKey: String {
        switch self {
        case .squad(let id):
            return "squad:\(id)"
        case .direct(let id):
            return "direct:\(id)"
        }
    }
}

struct ChatReadCursor: Codable, Equatable {
    var lastReadMessageId: String
    var lastReadAt: Date
}

final class ChatReadCursorStore {
    private static let schemaVersion = "v1"
    private static let trackingEpochSchemaVersion = "v1"
    private var userID: String?
    private var cursors: [String: ChatReadCursor] = [:]
    private(set) var trackingStartedAt: Date?

    func bind(userID rawUserID: String?) {
        let trimmed = rawUserID?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard trimmed != userID else { return }
        userID = trimmed.isEmpty ? nil : trimmed
        loadFromDisk()
        ensureTrackingEpochIfNeeded()
    }

    func cursor(for thread: ChatReadThreadKey) -> ChatReadCursor? {
        cursors[thread.storageKey]
    }

    func markRead(thread: ChatReadThreadKey, messageId: String, at readAt: Date) {
        let messageID = messageId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !messageID.isEmpty else { return }
        let key = thread.storageKey
        if let existing = cursors[key],
           existing.lastReadAt > readAt
           || (existing.lastReadAt == readAt && existing.lastReadMessageId >= messageID) {
            return
        }
        cursors[key] = ChatReadCursor(lastReadMessageId: messageID, lastReadAt: readAt)
        persistToDisk()
    }

    func clearAll() {
        cursors = [:]
        persistToDisk()
    }

    private func defaultsKey() -> String? {
        guard let userID, !userID.isEmpty else { return nil }
        return "otto.chatReadCursors.\(Self.schemaVersion).\(userID)"
    }

    private func loadFromDisk() {
        cursors = [:]
        trackingStartedAt = nil
        guard let userID, !userID.isEmpty else { return }
        if let epochKey = trackingEpochDefaultsKey(),
           let epoch = UserDefaults.standard.object(forKey: epochKey) as? Date {
            trackingStartedAt = epoch
        }
        guard let key = defaultsKey(),
              let data = UserDefaults.standard.data(forKey: key),
              let decoded = try? JSONDecoder().decode([String: ChatReadCursor].self, from: data) else {
            return
        }
        cursors = decoded
    }

    private func persistToDisk() {
        guard let key = defaultsKey(),
              let data = try? JSONEncoder().encode(cursors) else {
            return
        }
        UserDefaults.standard.set(data, forKey: key)
    }

    private func trackingEpochDefaultsKey() -> String? {
        guard let userID, !userID.isEmpty else { return nil }
        return "otto.chatUnreadTrackingStartedAt.\(Self.trackingEpochSchemaVersion).\(userID)"
    }

    private func ensureTrackingEpochIfNeeded() {
        guard let userID, !userID.isEmpty else {
            trackingStartedAt = nil
            return
        }
        if trackingStartedAt != nil { return }
        let now = Date()
        trackingStartedAt = now
        if let key = trackingEpochDefaultsKey() {
            UserDefaults.standard.set(now, forKey: key)
        }
    }
}

enum ChatUnreadCounting {
    static func isMessageUnread(
        messageID: String,
        createdAt: Date,
        senderUserID: String,
        isDeleted: Bool,
        cursor: ChatReadCursor?,
        currentUserID: String,
        trackingStartedAt: Date?
    ) -> Bool {
        guard !isDeleted else { return false }
        guard senderUserID != currentUserID else { return false }
        if let trackingStartedAt, createdAt <= trackingStartedAt { return false }
        guard let cursor else { return true }
        if createdAt > cursor.lastReadAt { return true }
        if createdAt < cursor.lastReadAt { return false }
        return messageID > cursor.lastReadMessageId
    }

    static func squadUnreadCount(
        messages: [CircleChatMessageDTO],
        cursor: ChatReadCursor?,
        currentUserID: String,
        trackingStartedAt: Date?
    ) -> Int {
        messages.reduce(into: 0) { count, message in
            if isMessageUnread(
                messageID: message.id,
                createdAt: message.createdAt,
                senderUserID: message.resolvedSenderUserId,
                isDeleted: message.deletedAt != nil,
                cursor: cursor,
                currentUserID: currentUserID,
                trackingStartedAt: trackingStartedAt
            ) {
                count += 1
            }
        }
    }

    static func directUnreadCount(
        messages: [DirectMessageDTO],
        cursor: ChatReadCursor?,
        currentUserID: String,
        trackingStartedAt: Date?
    ) -> Int {
        messages.reduce(into: 0) { count, message in
            if isMessageUnread(
                messageID: message.id,
                createdAt: message.createdAt,
                senderUserID: message.senderUserId,
                isDeleted: message.deletedAt != nil,
                cursor: cursor,
                currentUserID: currentUserID,
                trackingStartedAt: trackingStartedAt
            ) {
                count += 1
            }
        }
    }

    /// Minimum unread estimate from inbox preview when transcript cache is empty.
    static func previewUnreadEstimate(
        lastMessageAt: Date?,
        lastMessageSenderUserID: String?,
        cursor: ChatReadCursor?,
        currentUserID: String,
        trackingStartedAt: Date?
    ) -> Int {
        guard let lastMessageAt,
              let sender = lastMessageSenderUserID?.trimmingCharacters(in: .whitespacesAndNewlines),
              !sender.isEmpty,
              sender != currentUserID else {
            return 0
        }
        if let trackingStartedAt, lastMessageAt <= trackingStartedAt { return 0 }
        guard let cursor else { return 1 }
        if lastMessageAt > cursor.lastReadAt { return 1 }
        if lastMessageAt < cursor.lastReadAt { return 0 }
        return 1
    }
}

File: otto-android/app/src/main/java/to/ottomot/driftd/core/chat/ChatReadCursorStore.kt
package to.ottomot.driftd.core.chat

import android.content.Context
import androidx.core.content.edit
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.time.Instant

enum class ChatReadThreadKind {
    Squad,
    Direct,
}

data class ChatReadThreadKey(
    val kind: ChatReadThreadKind,
    val id: String,
) {
    val storageKey: String
        get() =
            when (kind) {
                ChatReadThreadKind.Squad -> "squad:$id"
                ChatReadThreadKind.Direct -> "direct:$id"
            }
}

data class ChatReadCursor(
    val lastReadMessageId: String,
    /** Epoch milliseconds for Gson persistence. */
    val lastReadAtEpochMs: Long,
) {
    val lastReadAt: Instant
        get() = Instant.ofEpochMilli(lastReadAtEpochMs)
}

class ChatReadCursorStore(
    context: Context,
    private val gson: Gson = Gson(),
) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private var userId: String? = null
    private var cursors: MutableMap<String, ChatReadCursor> = mutableMapOf()
    var trackingStartedAt: Instant? = null
        private set

    fun bind(userIdRaw: String?) {
        val trimmed = userIdRaw?.trim()?.takeIf { it.isNotEmpty() }
        if (trimmed == userId) return
        userId = trimmed
        loadFromDisk()
        ensureTrackingEpochIfNeeded()
    }

    fun cursor(thread: ChatReadThreadKey): ChatReadCursor? = cursors[thread.storageKey]

    fun markRead(thread: ChatReadThreadKey, messageId: String, readAt: Instant) {
        val messageID = messageId.trim().takeIf { it.isNotEmpty() } ?: return
        val key = thread.storageKey
        val existing = cursors[key]
        if (existing != null) {
            if (existing.lastReadAt.isAfter(readAt)) return
            if (existing.lastReadAt == readAt && existing.lastReadMessageId >= messageID) return
        }
        cursors[key] = ChatReadCursor(lastReadMessageId = messageID, lastReadAtEpochMs = readAt.toEpochMilli())
        persistToDisk()
    }

    fun clearAll() {
        cursors.clear()
        persistToDisk()
    }

    private fun prefsKey(): String? {
        val uid = userId?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        return "$PREFS_KEY_PREFIX.$SCHEMA_VERSION.$uid"
    }

    private fun loadFromDisk() {
        cursors.clear()
        trackingStartedAt = null
        val uid = userId?.trim()?.takeIf { it.isNotEmpty() } ?: return
        trackingEpochPrefsKey(uid)?.let { epochKey ->
            val epochMs = prefs.getLong(epochKey, Long.MIN_VALUE)
            if (epochMs != Long.MIN_VALUE) {
                trackingStartedAt = Instant.ofEpochMilli(epochMs)
            }
        }
        val key = prefsKey() ?: return
        val json = prefs.getString(key, null) ?: return
        val type = object : TypeToken<Map<String, ChatReadCursor>>() {}.type
        val decoded: Map<String, ChatReadCursor>? = runCatching { gson.fromJson<Map<String, ChatReadCursor>>(json, type) }.getOrNull()
        if (decoded != null) {
            cursors.putAll(decoded)
        }
    }

    private fun persistToDisk() {
        val key = prefsKey() ?: return
        val json = gson.toJson(cursors)
        prefs.edit { putString(key, json) }
    }

    private fun trackingEpochPrefsKey(uid: String): String? =
        "$TRACKING_EPOCH_KEY_PREFIX.$TRACKING_EPOCH_SCHEMA_VERSION.$uid"

    private fun ensureTrackingEpochIfNeeded() {
        val uid = userId?.trim()?.takeIf { it.isNotEmpty() } ?: run {
            trackingStartedAt = null
            return
        }
        if (trackingStartedAt != null) return
        val now = Instant.now()
        trackingStartedAt = now
        trackingEpochPrefsKey(uid)?.let { key ->
            prefs.edit { putLong(key, now.toEpochMilli()) }
        }
    }

    companion object {
        private const val PREFS_NAME = "otto_chat_read_cursors"
        private const val PREFS_KEY_PREFIX = "otto.chatReadCursors"
        private const val SCHEMA_VERSION = "v1"
        private const val TRACKING_EPOCH_KEY_PREFIX = "otto.chatUnreadTrackingStartedAt"
        private const val TRACKING_EPOCH_SCHEMA_VERSION = "v1"
    }
}
